File: Stacks Queues/stacks/stack.js
class Stack {
    constructor() {
        this.stack = [];
        this.top = -1;
    }

    push(value) {
        this.stack[++this.top] = value;
        console.log(`Pushed ${value} onto the stack`);
    }

    pop() {
        if (this.top < 0) {
            console.log("Stack Underflow");
            return -1;
        }
        const value = this.stack[this.top--];
        console.log(`Popped ${value} from the stack`);
        return value;
    }

    peek() {
        if (this.top < 0) {
            console.log("Stack is empty");
            return -1;
        }
        return this.stack[this.top];
    }

    isEmpty() {
        return this.top === -1;
    }

    printStack() {
        if (this.isEmpty()) {
            console.log("Stack is empty");
            return;
        }
        console.log("Elements in the stack:");
        for (let i = this.top; i >= 0; i--) {
            console.log(this.stack[i]);
        }
    }
}

const stack = new Stack();
let choice, value;

const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

function menu() {
    console.log("Stack Menu:");
    console.log("1. Push");
    console.log("2. Pop");
    console.log("3. Peek");
    console.log("4. Check if stack is empty");
    console.log("5. Print stack");
    console.log("6. Exit");
    readline.question("Enter your choice: ", option => {
        choice = parseInt(option);
        switch (choice) {
            case 1:
                readline.question("Enter value to push: ", val => {
                    value = parseInt(val);
                    stack.push(value);
                    menu();
                });
                break;
            case 2:
                stack.pop();
                menu();
                break;
            case 3:
                console.log(`Top element: ${stack.peek()}`);
                menu();
                break;
            case 4:
                console.log(`Is stack empty? ${stack.isEmpty() ? "Yes" : "No"}`);
                menu();
                break;
            case 5:
                stack.printStack();
                menu();
                break;
            case 6:
                console.log("Exiting...");
                readline.close();
                break;
            default:
                console.log("Invalid choice");
                menu();
        }
    });
}

menu();

File: Stacks Queues/stacks/stack.py
class Stack:
    def __init__(self, limit):
        self.limit = limit
        self.stack = []
    
    def is_full(self):
        return len(self.stack) == self.limit
    
    def push(self, value):
        if self.is_full():
            print("Stack is full")
            return
        self.stack.append(value)
        print(f"Pushed {value} onto the stack")
    
    def pop(self):
        if not self.stack:
            print("Stack Underflow")
            return -1
        value = self.stack.pop()
        print(f"Popped {value} from the stack")
        return value
    
    def peek(self):
        if not self.stack:
            print("Stack is empty")
            return -1
        return self.stack[-1]
    
    def is_empty(self):
        return not self.stack
    
    def print_stack(self):
        if self.is_empty():
            print("Stack is empty")
            return
        print("Elements in the stack:")
        for value in reversed(self.stack):
            print(value)

def main():
    limit = int(input("Enter the maximum number of elements for the stack: "))
    stack = Stack(limit)
    
    while True:
        print("Stack Menu:")
        print("1. Push")
        print("2. Pop")
        print("3. Peek")
        print("4. Check if stack is empty")
        print("5. Print stack")
        print("6. Exit")
        choice = int(input("Enter your choice: "))
        
        if choice == 1:
            if stack.is_full():
                print("Stack is full")
            else:
                value = int(input("Enter value to push: "))
                stack.push(value)
        elif choice == 2:
            stack.pop()
        elif choice == 3:
            print(f"Top element: {stack.peek()}")
        elif choice == 4:
            print(f"Is stack empty? {'Yes' if stack.is_empty() else 'No'}")
        elif choice == 5:
            stack.print_stack()
        elif choice == 6:
            print("Exiting...")
            break
        else:
            print("Invalid choice")

if __name__ == "__main__":
    main()
